rust/hyperlane-base/src/contract_sync/cursors/sequence_aware/forward.rs

```diff
@@ -0,0 +1,1089 @@
+        };
+
+        let current_sequence = self.current_indexing_snapshot.sequence;
+        let range = match current_sequence.cmp(&onchain_sequence_count) {
+            Ordering::Equal => {
+                // We are synced up to the latest sequence so we don't need to index anything.
+
+                // We can update the current indexing snapshot to the tip.
+                // This will let us only index blocks that are likely to have new logs once
+                // there's a new sequence to search for.
+                self.current_indexing_snapshot.at_block = tip;
+
+                None
+            }
+            Ordering::Less => {
+                // The cursor is behind the onchain sequence count, so we need to index.
+
+                // Minus one because this is the sequence we're targeting, not the count.
+                let target_sequence = onchain_sequence_count.saturating_sub(1);
+
+                // Set the target to the highest sequence and tip.
+                // We don't necessarily expect to hit this target in the next query (because we
+                // have limits to the range size based off the chunk size), but we will use it
+                // as an eventual target.
+                self.target_snapshot = Some(TargetSnapshot {
+                    sequence: target_sequence,
+                    at_block: tip,
+                });
+
+                match &self.index_mode {
+                    IndexMode::Block => {
+                        // Query the block range starting from the current_indexing_snapshot's at_block.
+                        Some(
+                            self.current_indexing_snapshot.at_block
+                                ..=u32::min(
+                                    self.current_indexing_snapshot.at_block + self.chunk_size,
+                                    tip,
+                                ),
+                        )
+                    }
+                    IndexMode::Sequence => {
+                        // Query the sequence range starting from the cursor count.
+                        Some(
+                            current_sequence
+                                ..=u32::min(target_sequence, current_sequence + self.chunk_size),
+                        )
+                    }
+                }
+            }
+            Ordering::Greater => {
+                // Providers may be internally inconsistent, e.g. RPC request A could hit a node
+                // whose tip is N and subsequent RPC request B could hit a node whose tip is < N.
+                // Just warn and try to continue as normal.
+                warn!(
+                    current_sequence,
+                    onchain_sequence_count,
+                    "Current sequence is greater than the onchain sequence count"
+                );
+                None
+            }
+        };
+
+        Ok(range)
+    }
+
+    async fn fast_forward(&mut self) -> Result<()> {
```

Comment:
I'd add a doc comment to this function to explain that fast-forwarding is possible due to the local db, and how that's done